### PR TITLE
Add UI support for displaying cluster badges next to cluster name

### PR DESCRIPTION
**Pull request by neillsom**
#4934 

This PR adds ui/ux functionality in support of PR #4933

### CODE REVIEW

File: config/labels-annotations.js [hunk 1]
```diff
@@ -157,5 +157,5 @@ export const CLUSTER_BADGE = {
   // Badge color - as a hex color - e.g. #ff00ff
   COLOR:       'ui.rancher/badge-color',
   // Use as icon - if set to "true" the cluster icon will also reflect the badge
-  ICON_TEXT: 'ui.rancher/badge-use-as-icon',
```
Comment (nwmac) [inline config/labels-annotations.js:160]: This undoes the change to allow the icon text to be specified

File: models/management.cattle.io.cluster.js [hunk 2]
```diff
@@ -237,13 +237,15 @@ export default class MgmtCluster extends HybridModel {
...
-    const iconText = this.metadata?.annotations[CLUSTER_BADGE.ICON_TEXT] || '';
+    const useForIcon = this.metadata?.annotations[CLUSTER_BADGE.USE_AS_ICON] || 'false';
+    const letter = this.metadata?.annotations[CLUSTER_BADGE.TEXT].substr(0, 1).toUpperCase() || '';
 
     return {
       text,
       color,
       textColor: textColor(parseColor(color)),
-      iconText:  iconText.substr(0, 2),
+      useForIcon,
```
Comment (nwmac) [inline models/management.cattle.io.cluster.js:247]: This undoes the change to allow the icon text to be specified

File: pages/c/_cluster/explorer/index.vue [hunk 3]
```diff
@@ -359,6 +367,18 @@ export default {
         <span><LiveDate :value="currentCluster.metadata.creationTimestamp" :add-suffix="true" :show-tooltip="true" /></span>
       </div>
       <div :style="{'flex':1}" />
+      <!--  -->
+      <div>
+        <a
+          class="badge-install"
+          @click="customBadgeDialog"
+        >
+          <i class="icon icon-star" />
```
Comment (nwmac) [inline pages/c/_cluster/explorer/index.vue:376]: Ask Lauren for an icon for a cluster badge

**Review by nwmac — CHANGES_REQUESTED** [review]
A few things:

- Need to check if the user has permission to edit the cluster and use this to determine if should show add/edit badge
- Can we move the badge add/edit after the monitoring link
- Can we change the dialog so that instead of hiding the contents when you change the checkbox, it disables the controls - this will stop the dialog from changing height when you do this
- Can you update to support specifying the icon text (max 2 letters) instead of just the checkbox - Suggest you have a checkbox 'Customise cluster icon' which when checked allows you to specify 1 or 2 chars to use as the text in the icon.

File: pages/c/_cluster/explorer/index.vue (hunk 3 above)
Comment (lvuch) [reply pages/c/_cluster/explorer/index.vue:376]: I believe icon-cluster would work. Unless you want something more specific that illustrates a badge.

File: config/labels-annotations.js [hunk 4]
```diff
@@ -157,5 +157,7 @@ export const CLUSTER_BADGE = {
...
+  USE_AS_ICON: 'ui.rancher/badge-use-as-icon',
```
Comment (nwmac) [inline config/labels-annotations.js:160]: the `USE_AS_ICON` annotation should go - we infer this from the presence of the `ICON_TEXT` annotation (when it has 1 or 2 chars)

File: models/management.cattle.io.cluster.js [hunk 5]
```diff
@@ -236,14 +236,16 @@ export default class MgmtCluster extends HybridModel {
       return undefined;
     }
 
-    const color = this.metadata?.annotations[CLUSTER_BADGE.COLOR] || '#7f7f7f';
-    const iconText = this.metadata?.annotations[CLUSTER_BADGE.ICON_TEXT] || '';
+    const color = this.metadata?.annotations[CLUSTER_BADGE.COLOR];
```
Comment (nwmac) [inline models/management.cattle.io.cluster.js:239]: We've lost the default color #7f7f7f - this addresses a corner case

File: models/management.cattle.io.cluster.js [hunk 6]
```diff
@@ -236,14 +236,16 @@ export default class MgmtCluster extends HybridModel {
...
+    const useForIcon = this.metadata?.annotations[CLUSTER_BADGE.USE_AS_ICON] || '';
+    const letter = this.metadata?.annotations[CLUSTER_BADGE.ICON_TEXT] || text.substr(0, 1).toUpperCase();
```
Comment (nwmac) [inline models/management.cattle.io.cluster.js:241]: This only take first char - we support up to 2 - which is what the `iconText` was doing before (see line 246)

File: components/dialog/AddCustomBadgeDialog.vue [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+        delete norman.annotations[CLUSTER_BADGE.TEXT];
+        delete norman.annotations[CLUSTER_BADGE.COLOR];
+        delete norman.annotations[CLUSTER_BADGE.USE_AS_ICON];
+        delete norman.annotations[CLUSTER_BADGE.ICON_TEXT];
+
+        if (this.useCustomBadge) {
+          this.$set(norman.annotations, CLUSTER_BADGE.TEXT, this.badgeDescription);
+          this.$set(norman.annotations, CLUSTER_BADGE.COLOR, this.badgeBgColor);
+          this.$set(norman.annotations, CLUSTER_BADGE.USE_AS_ICON, this.badgeAsIcon ? 'true' : 'false');
+          this.$set(norman.annotations, CLUSTER_BADGE.ICON_TEXT, this.letter.toUpperCase());
```
Comment (nwmac) [inline components/dialog/AddCustomBadgeDialog.vue:104]: Should support up to 12 chars - see further down

File: components/dialog/AddCustomBadgeDialog.vue [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+
+      this.letter = this.currentCluster.metadata?.annotations[CLUSTER_BADGE.ICON_TEXT] || this.badgeDescription.substring(0, 1);
+    }
+  },
+
+  computed: {
+    ...mapGetters(['currentCluster']),
+
+    previewColor() {
+      return textColor(parseColor(this.badgeBgColor)) || 'white';
```
Comment (nwmac) [inline components/dialog/AddCustomBadgeDialog.vue:59]: prefer to use # color codes - #fff

File: components/dialog/AddCustomBadgeDialog.vue [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+    previewCluster() {
+      // Make cluster object that is enough for the badge component to work
+      return {
+        isLocal:         this.currentCluster.isLocal,
+        providerNavLogo: this.currentCluster.providerNavLogo,
+        badge:           {
+          text:        this.badgeDescription,
+          color:       this.badgeBgColor,
+          textColor:   textColor(parseColor(this.badgeBgColor)),
+          useForIcon:  this.badgeAsIcon ? 'true' : 'false',
```
Comment (nwmac) [inline components/dialog/AddCustomBadgeDialog.vue:74]: infer this from the presence of the icon text

**Review by nwmac — CHANGES_REQUESTED** [review]

**Comment by nwmac** [comment]
@neillsom Merging on your behalf

**Pull request merged**